src/platform/webusb/device.rs

```diff
@@ -0,0 +1,223 @@
+    descriptors::{validate_config_descriptor, DESCRIPTOR_TYPE_CONFIGURATION},
+    transfer::{Control, EndpointType, TransferError, TransferHandle},
+    DeviceInfo, Error,
+};
+
+pub(crate) struct WebusbDevice {
+    pub device: Arc<UsbDevice>,
+    config_descriptors: Vec<Vec<u8>>,
+}
+
+/// SAFETY: This is NOT safe at all.
+unsafe impl Sync for WebusbDevice {}
+unsafe impl Send for WebusbDevice {}
+
+impl WebusbDevice {
+    pub(crate) async fn from_device_info(d: &DeviceInfo) -> Result<Arc<WebusbDevice>, Error> {
+        let window = web_sys::window().unwrap();
+        let navigator = window.navigator();
+        let usb = navigator.usb();
+        let devices = JsFuture::from(usb.get_devices()).await.unwrap();
+
+        let devices: Array = JsCast::unchecked_from_js(devices);
+
+        for device in devices {
+            let device: UsbDevice = JsCast::unchecked_from_js(device);
+            let device = Arc::new(device);
+            if device.vendor_id() == d.vendor_id
+                && device.product_id() == d.product_id
+                && device.serial_number() == d.serial_number
+            {
+                JsFuture::from(device.open()).await.unwrap();
+
+                let config_descriptors = extract_decriptors(&device).await;
+
+                return Ok(Arc::new(Self {
+                    device,
+                    config_descriptors,
+                }));
+            }
+        }
+        Err(Error::other("device not found"))
+    }
+
+    // pub(crate) fn handle_events(&self) {
+    //     todo!()
+    // }
+
+    pub(crate) fn configuration_descriptors(&self) -> impl Iterator<Item = &[u8]> {
+        self.config_descriptors.iter().map(|d| &d[..])
+    }
+
+    pub(crate) fn active_configuration_value(&self) -> u8 {
+        todo!()
+    }
+
+    pub(crate) fn set_configuration(&self, _configuration: u8) -> Result<(), Error> {
+        todo!()
+    }
+
+    pub(crate) fn reset(&self) -> Result<(), Error> {
+        todo!()
+    }
+
+    /// SAFETY: `data` must be valid for `len` bytes to read or write, depending on `Direction`
+    // unsafe fn control_blocking(
+    //     &self,
+    //     _direction: Direction,
+    //     _control: Control,
+    //     _data: *mut u8,
+    //     _len: usize,
+    //     _timeout: Duration,
+    // ) -> Result<usize, TransferError> {
+    //     todo!()
+    // }
+
+    // pub fn control_in_blocking(
+    //     &self,
+    //     _control: Control,
+    //     _data: &mut [u8],
+    //     _timeout: Duration,
+    // ) -> Result<usize, TransferError> {
+    //     todo!()
+    // }
+
+    // pub fn control_out_blocking(
+    //     &self,
+    //     _control: Control,
+    //     _data: &[u8],
+    //     _timeout: Duration,
+    // ) -> Result<usize, TransferError> {
+    //     todo!()
+    // }
+
+    // pub(crate) fn make_control_transfer(self: &Arc<Self>) -> TransferHandle<super::TransferData> {
+    //     todo!()
+    // }
+
+    pub(crate) async fn claim_interface(
```

Comment:
Is that a problem? Or do you think we can get away with it? I can maybe move the future to a thread if it blocks too long (tokio does not like that but I also don't want to assume tokio).

Reply:
This applies to `Device::open`, `Device::claim_interface`, `Device::set_configuration`, `Device::reset`, `Interface::set_alt_setting`, `Interface::clear_halt`.

Some of these operations like `set_configuration`, `set_alt_setting`, and `clear_halt` involve control transfers to the device, so can wait arbitrarily long and really should be async. The problem is that the OS APIs for all of these operations are blocking. As you mention, the "right" way to handle that is using something that runs the blocking operations as a threadpool, like [smol's `blocking`](https://github.com/smol-rs/blocking) or [tokio `spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). Even without wasm, there could be nusb users that want that to avoid blocking their event loops.

However, the question then is which runtime, and I would like to make it possible to use nusb without needing to pull in either big async runtime for (non-wasm) users who only want blocking operations anyway. Could do Cargo feature flags, but that's a maintenance burden.

The other case is WebUSB operations like `release_interface` and `close` that are triggered by dropping nusb types, since Rust doesn't have async drop yet. Since `Interface` and `Device` are ref-counted and already wait for transfers to cancel, dropping one already doesn't guarantee that anything is closed right away, so I don't expect this to be a problem. Could offer an async `release` method consuming self that waits for drop or errors if other references exist, if you for some reason want to be able to release an interface and then claim it again right away.

Reply:
Trying this out in https://github.com/kevinmehall/nusb/pull/100. The idea is that `IoAction::wait()` would be cfg'd out on wasm, and then you could trivially `impl<T> IoAction for T where T: Future`.